README.md
# Stormpath is Joining Okta
We are incredibly excited to announce that [Stormpath is joining forces with Okta](https://stormpath.com/blog/stormpaths-new-path?utm_source=github&utm_medium=readme&utm-campaign=okta-announcement). Please visit [the Migration FAQs](https://stormpath.com/oktaplusstormpath?utm_source=github&utm_medium=readme&utm-campaign=okta-announcement) for a detailed look at what this means for Stormpath users.

We're available to answer all questions at [support@stormpath.com](mailto:support@stormpath.com).

# Stormpath Angular SDK
[![Build Status](https://travis-ci.org/stormpath/stormpath-sdk-angular.svg?branch=master)](https://travis-ci.org/stormpath/stormpath-sdk-angular)
[![npm version](https://badge.fury.io/js/angular-stormpath.svg)](http://badge.fury.io/js/angular-stormpath)
[![devDependency Status](https://david-dm.org/stormpath/stormpath-sdk-angular/dev-status.svg)](https://david-dm.org/stormpath/stormpath-sdk-angular#info=devDependencies)
[![GitHub issues](https://img.shields.io/github/issues/stormpath/stormpath-sdk-angular.svg)](https://github.com/stormpath/stormpath-sdk-angular/issues)
[![GitHub stars](https://img.shields.io/github/stars/stormpath/stormpath-sdk-angular.svg)](https://github.com/stormpath/stormpath-sdk-angular/stargazers)
[![GitHub license](https://img.shields.io/badge/license-APACHE-red.svg)](https://raw.githubusercontent.com/stormpath/stormpath-sdk-angular/master/LICENSE)

> Angular Components for integrating with Stormpath's API

<div>
  <a href="http://angular.io">
    <img src="https://github.com/stormpath/stormpath-sdk-angular/raw/master/static/angular.png"></a>
  <a href="https://www.stormpath.com/">
    <img src="https://github.com/stormpath/stormpath-sdk-angular/raw/master/static/stormpath.png"></a>
</div>

## Table of contents

- [About](#about)
- [Installation](#installation)
  - [Configuration](#configuration)
- [Documentation](#documentation)
- [Development](#development)
- [License](#licence)

## About

Angular SDK for Stormpath's API. If you're looking for **AngularJS** support, please see [stormpath-sdk-angularjs](https://github.com/stormpath/stormpath-sdk-angularjs).

## Installation

Install through npm:
```
npm install --save angular-stormpath
```

Then use it in your app like so:

```typescript
import { Component } from '@angular/core';
import { AuthPortComponent } from 'angular-stormpath';

@Component({
  selector: 'app-root',
  template: `<div *ngIf="(user$ | async)" class="row text-center">
 <h2>
   Welcome, {{ ( user$ | async ).fullName }}.
 </h2>

 <ul class="nav nav-pills nav-stacked text-centered">
   <li role="presentation" (click)="logout()"><a href="#">Logout</a></li>
 </ul>
</div>

<sp-authport></sp-authport>`
})
export class AppComponent extends AuthPortComponent {
}
```

If you're using a version of Angular < 2.3, extending AuthPortComponent won't work for you. As an alternative, you can inject the `Stormpath` service into your component, subscribe to `stormpath.user$` and implement `logout()` yourself.

```typescript
import { Account, Stormpath } from 'angular-stormpath';
...
export class AppComponent {
  user$: Observable<Account | boolean>;

  constructor(private stormpath: Stormpath) {
    this.user$ = this.stormpath.user$;
  }

  logout(): void {
    this.stormpath.logout();
  }
}
```

You may also find it useful to view the [demo source](https://github.com/stormpath/stormpath-sdk-angular/blob/master/demo/app.component.ts).

### Configuration

To override the endpoint prefix or URIs for the various endpoints, you can modify the defaults in [StormpathConfiguration](https://github.com/stormpath/stormpath-sdk-angular/blob/master/src/stormpath/stormpath.config.ts).

For example, to override the endpoint prefix and `/me` URI in [demo.module.ts](https://github.com/stormpath/stormpath-sdk-angular/blob/master/demo/demo.module.ts), change it to the following:

```typescript
export function stormpathConfig(): StormpathConfiguration {
 let spConfig: StormpathConfiguration = new StormpathConfiguration();
 spConfig.endpointPrefix = 'http://api.mycompany.com';
 spConfig.meUri = '/account';
 return spConfig;
}

@NgModule({
  declarations: [AppComponent],
  imports: [BrowserModule, StormpathModule],
  bootstrap: [AppComponent],
  providers: [{
    provide: StormpathConfiguration, useFactory: stormpathConfig
  }]
})
export class DemoModule {
}
```

#### OAuth

If your Angular app is on a different domain than your endpoints, OAuth will be used for login/logout. The access token will be stored in localStorage under the name `stormpath:token` and it will be automatically added as an `Authorization` header when you send HTTP requests to your `/me` endpoint. 

If you'd like to add this header to additional URLs, you'll need to add them as follows:

```typescript
let config: StormpathConfiguration = new StormpathConfiguration();
config.autoAuthorizedUris.push(new RegExp('http://localhost:3000/myapi/*)');
```

#### Templates

To override templates, you can use the `customTemplate` attribute on a component. Below is an example of [app.component.ts](https://github.com/stormpath/stormpath-sdk-angular/blob/master/demo/app.component.ts) with a custom `<sp-authport>` and `<login-form>`:

```typescript
import { Component, OnInit } from '@angular/core';
import { Observable } from 'rxjs/Observable';
import { Stormpath, StormpathErrorResponse, Account, LoginFormModel } from 'angular-stormpath';

@Component({
  selector: 'demo-app',
  template: `<div class="container">
<div *ngIf="(user$ | async)" class="row text-center">
  <h2>
    Welcome, {{ ( user$ | async ).fullName }}.
  </h2>
  <hr/>

  <ul class="nav nav-pills nav-stacked text-centered">
    <li role="presentation" (click)="logout(); false"><a href="">Logout</a></li>
  </ul>
</div>

<ng-template #loginform>
  <div *ngIf="error" class="alert alert-danger">{{error}}</div>
  <form>
      <label for="email">Email</label>
      <input id="email" name="login" type="text" [(ngModel)]="loginFormModel.login">
      <label for="passwordField">Password</label>
      <input id="passwordField" name="password" type="password" [(ngModel)]="loginFormModel.password">
      <button (click)="login()">Login</button>
  </form>
</ng-template>

<ng-template #authport>
  <div *ngIf="(user$ | async) === false">
      <h2>Sign In</h2>
      <login-form [customTemplate]="loginform"></login-form>
  </div>
</ng-template>

<sp-authport [customTemplate]="authport"></sp-authport>

</div>`
})
export class AppComponent implements OnInit {

  protected loginFormModel: LoginFormModel;
  protected error: string;

  private user$: Observable<Account | boolean>;
  private loggedIn$: Observable<boolean>;
  private _login: boolean;
  private _register: boolean;

  constructor(public stormpath: Stormpath) {
    this.loginFormModel = {
      login: '',
      password: ''
    };
  }

  ngOnInit(): void {
    this._login = true;
    this._register = false;
    this.user$ = this.stormpath.user$;
    this.loggedIn$ = this.user$.map(user => !!user);
  }

  showLogin(): void {
    this._login = !(this._register = false);
  }

  showRegister(): void {
    this._register = !(this._login = false);
  }

  login(): void {
    this.error = null;
    this.stormpath.login(this.loginFormModel)
      .subscribe(null, (error: StormpathErrorResponse) => {
        this.error = error.message;
      });
  }

  logout(): void {
    this.stormpath.logout();
  }
}
```

**NOTE:** One problem with this approach is you'll need to copy all the referenced variables in the template into your component. Another option is to extend the existing Stormpath component and override its `template` variable in `@Component`. This is the recommended solution if you're using Angular 2.3+.

#### Access Token Storage

To change the storage mechanism for authentication tokens from localStorage (the default), to cookies, change the class for the 'tokenStore' provider.

```typescript
{
  provide: 'tokenStore', useClass: CookieTokenStoreManager
}
```

Below is a list of direct links to each component. You can use the HTML defined in their `template` variable as a starting point for your customizations.

* [authport.component.ts](https://github.com/stormpath/stormpath-sdk-angular/blob/master/src/authport/authport.component.ts)
* [email-verification.component.ts](https://github.com/stormpath/stormpath-sdk-angular/blob/master/src/email-verification/email-verification.component.ts)
* [forgot-password.component.ts](https://github.com/stormpath/stormpath-sdk-angular/blob/master/src/forgot-password/forgot-password.component.ts)
* [login.component.ts](https://github.com/stormpath/stormpath-sdk-angular/blob/master/src/login/login.component.ts)
* [register.component.ts](https://github.com/stormpath/stormpath-sdk-angular/blob/master/src/register/register.component.ts)
* [resend-email-verification.component.ts](https://github.com/stormpath/stormpath-sdk-angular/blob/master/src/resend-email-verification/resend-email-verification.component.ts)
* [reset-password-component.component.ts](https://github.com/stormpath/stormpath-sdk-angular/blob/master/src/reset-password/reset-password.component.ts)

### Usage without a module bundler
```
<script src="node_modules/dist/umd/stormpath-sdk-angular/stormpath-sdk-angular.js"></script>
<script>
    // everything is exported Stormpath namespace
</script>
```

## Documentation
All documentation is auto-generated from the source via typedoc and can be viewed at https://docs.stormpath.com/angular/api.

## Development

### Prepare your environment
* Install [Node.js](http://nodejs.org/) and npm
* Install local dev dependencies: `npm install` while current directory is this repo

### Development server
Run `npm start` to start a development server on port 8000 with auto reload + tests.

### Testing
Run `npm test` to run tests once or `npm run test:watch` to continually run tests.

### Using npm link

If you want to use `npm link` to use this module in another Angular project, follow the steps below:

1. Build this project using `npm run build:dist`.
2. Run `npm link` in this project's directory.
3. Run `npm link angular-stormpath` in the `<test-project>`.
4. Run `rm -rf node_modules/angular-stormpath/node_modules` in `<test-project>`.
5. Manually install dependencies required by `angular-stormpath`:

```
npm install ng2-webstorage angular2-cookie --save
```

You'll know if this doesn't work because you'll see an error message like the following:

```
Type 'Observable<boolean | Account>' is not assignable to type 'Observable<boolean | Account>'. '
Property 'source' is protected but type 'Observable<T>' is not a class derived from 'Observable<T>'.
```

When this happens, you can use the ol' fashioned copy and paste method. If you have `stormpath-sdk-angular` in an adjacent directory, the commands below should work on your project that has `angular-stormpath` already installed. You will need to run `npm run build:dist` every time you change code in this project.

```
rm -rf node_modules/angular-stormpath/dist
cp -r ../stormpath-sdk-angular/dist node_modules/angular-stormpath/.
cp ../stormpath-sdk-angular/package.json node_modules/angular-stormpath/.
```

### Release

Bump the version in package.json (once the module hits 1.0 this will become automatic).

```bash
npm run release
```

For more information, see [generator-angular-library](https://www.npmjs.com/package/generator-angular-library). 
It was used to create this project.

## License

Apache-2.0 © [Stormpath](https://stormpath.com)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.2.0"></a>
# [0.2.0](https://github.com/stormpath/stormpath-sdk-angular/compare/v0.1.7...v0.2.0) (2017-11-21)



<a name="2.0.0"></a>
## [0.2.0](https://github.com/stormpath/stormpath-sdk-angular/compare/v0.1.6...v0.2.0) (2017-11-21)


### Bug Fixes

* **package.json:** Remove angular-cookie and webstorage from devDependencies ([833acbb](https://github.com/stormpath/stormpath-sdk-angular/commit/833acbb))



<a name="0.1.6"></a>
## [0.1.6](https://github.com/stormpath/stormpath-sdk-angular/compare/v0.1.5...v0.1.6) (2017-02-04)


### Bug Fixes

* **dependencies:** Remove Ionic pages and module ([a3e837b](https://github.com/stormpath/stormpath-sdk-angular/commit/a3e837b))



<a name="0.1.5"></a>
## [0.1.5](https://github.com/stormpath/stormpath-sdk-angular/compare/v0.1.4...v0.1.5) (2017-02-03)


### Features

* **templates:** Add pages for Ionic with Ionic components in templates ([4360bab](https://github.com/stormpath/stormpath-sdk-angular/commit/4360bab))



<a name="0.1.4"></a>
## [0.1.4](https://github.com/stormpath/stormpath-sdk-angular/compare/v0.1.3...v0.1.4) (2017-01-30)


### Bug Fixes

* **build:** Move ng2-webstorage and angular2-cookies to dependencies since they're required ([d05d730](https://github.com/stormpath/stormpath-sdk-angular/commit/d05d730)), closes [#62](https://github.com/stormpath/stormpath-sdk-angular/issues/62)



<a name="0.1.3"></a>
## [0.1.3](https://github.com/stormpath/stormpath-sdk-angular/compare/v0.1.2...v0.1.3) (2017-01-30)


### Bug Fixes

* **angular:** Change to peerDependencies and allow Angular 2.0.0+ ([cbf9637](https://github.com/stormpath/stormpath-sdk-angular/commit/cbf9637))



<a name="0.1.2"></a>
## [0.1.2](https://github.com/stormpath/stormpath-sdk-angular/compare/v0.1.1...v0.1.2) (2017-01-27)


### Bug Fixes

* **http:** Fix compatibility with Angular versions < 2.3 ([b3ae345](https://github.com/stormpath/stormpath-sdk-angular/commit/b3ae345))
* **logout:** Fix logout fails when no access token found ([6fcaeda](https://github.com/stormpath/stormpath-sdk-angular/commit/6fcaeda))



<a name="0.1.1"></a>
## [0.1.1](https://github.com/stormpath/stormpath-sdk-angular/compare/v0.1.0...v0.1.1) (2017-01-20)


### Bug Fixes

* **config:** Allow autoAuthorizedUris to be modified ([#48](https://github.com/stormpath/stormpath-sdk-angular/issues/48)) ([ca81cbe](https://github.com/stormpath/stormpath-sdk-angular/commit/ca81cbe)), closes [#47](https://github.com/stormpath/stormpath-sdk-angular/issues/47)
* **dependencies:** Lock down version of ng2-webstorage to 1.4.3. ([6b00209](https://github.com/stormpath/stormpath-sdk-angular/commit/6b00209)), closes [#46](https://github.com/stormpath/stormpath-sdk-angular/issues/46)
* **security:** Set autocomplete="off" for all forms ([#52](https://github.com/stormpath/stormpath-sdk-angular/issues/52)) ([e85dc7d](https://github.com/stormpath/stormpath-sdk-angular/commit/e85dc7d)), closes [#51](https://github.com/stormpath/stormpath-sdk-angular/issues/51)



<a name="0.1.0"></a>
## [0.1.0](https://github.com/stormpath/stormpath-sdk-angular/compare/v0.0.5...v0.1.0) (2017-01-12)


### Bug Fixes

* **build:** Fix installation by downgrading to TypeScript 2.0.10. ([#45](https://github.com/stormpath/stormpath-sdk-angular/issues/45)) ([2ec5ae3](https://github.com/stormpath/stormpath-sdk-angular/commit/2ec5ae3))
* **demo:** Fix logout link so event propogation is stopped ([7758ed2](https://github.com/stormpath/stormpath-sdk-angular/commit/7758ed2)), closes [#22](https://github.com/stormpath/stormpath-sdk-angular/issues/22)
* **forms:** Add required field validation to login and forgot password forms. ([#40](https://github.com/stormpath/stormpath-sdk-angular/issues/40)) ([ca2f339](https://github.com/stormpath/stormpath-sdk-angular/commit/ca2f339)), closes [#24](https://github.com/stormpath/stormpath-sdk-angular/issues/24)
* **webpack:** Change X-Stormpath-Agent to read versions from package.json and Angular ([8e3747e](https://github.com/stormpath/stormpath-sdk-angular/commit/8e3747e))


### Features

* **oauth:** Add support for OAuth and Client API ([#37](https://github.com/stormpath/stormpath-sdk-angular/issues/37)) ([e18ed49](https://github.com/stormpath/stormpath-sdk-angular/commit/e18ed49))
* **tests:** Add Protractor configuration and tests ([41c216a](https://github.com/stormpath/stormpath-sdk-angular/commit/41c216a))


### Performance Improvements

* **webpack:** Upgrade to webpack 2 for better performance and ES6 ([9c8b345](https://github.com/stormpath/stormpath-sdk-angular/commit/9c8b345))



<a name="0.0.5"></a>
## [0.0.5](https://github.com/stormpath/stormpath-sdk-angular/compare/v0.0.4...v0.0.5) (2016-12-14)


### Bug Fixes

* **cors:** Use withCredentials: true for requests so CORS works ([991af7b](https://github.com/stormpath/stormpath-sdk-angular/commit/991af7b))


<a name="0.0.4"></a>
## [0.0.4](https://github.com/stormpath/stormpath-sdk-angular/compare/v0.0.3...v0.0.4) (2016-12-06)


### Bug Fixes

* resolving conflicted typedoc dependency ([da6fa97](https://github.com/stormpath/stormpath-sdk-angular/commit/da6fa97))
* **templates:** Fixed 404 in templates by embedding them ([cd0d377](https://github.com/stormpath/stormpath-sdk-angular/commit/cd0d377))
* **typedoc:** Fix typedoc listing of classes in right column ([0a464a4](https://github.com/stormpath/stormpath-sdk-angular/commit/0a464a4))



<a name="0.0.4"></a>
## [0.0.4](https://github.com/stormpath/stormpath-sdk-angular/compare/v0.0.3...v0.0.4) (2016-12-06)


### Bug Fixes

* **templates:** Fixed 404 in templates by embedding them ([98982c4](https://github.com/stormpath/stormpath-sdk-angular/commit/98982c4))
* **typedoc:** Fix typedoc listing of classes in right column ([0a464a4](https://github.com/stormpath/stormpath-sdk-angular/commit/0a464a4))


webpack.config.js
const path = require('path');
const webpack = require('webpack');
const IS_PROD = process.argv.indexOf('-p') > -1;
const LoaderOptionsPlugin = require("webpack/lib/LoaderOptionsPlugin");
const StringReplacePlugin = require('string-replace-webpack-plugin');
const TsConfigPathsPlugin = require('awesome-typescript-loader').TsConfigPathsPlugin;
const TOKENS = {
  VERSION: JSON.stringify(require('./package.json').version).replace(/['"]+/g, '')
};
module.exports = {
  devtool: IS_PROD ? 'source-map' : 'eval',
  entry: './demo/entry.ts',
  output: {
    filename: 'demo.js',
    path: IS_PROD ? './demo' : '/'
  },
  module: {
    rules: [
      {
        test: /\.ts$/,
        enforce: 'pre',
        loader: 'tslint-loader',
        exclude: /node_modules/
      },
      {
        test: /\.ts$/,
        loaders: ['awesome-typescript-loader', 'angular2-template-loader?keepUrl=true'],
        exclude: [/\.(spec|e2e)\.ts$/, /node_modules/]
      },
      /* Replace version from package.json */
      {
        test: /\.config.ts$/,
        loader: StringReplacePlugin.replace({
          replacements: [{
            pattern: /\${(.*)}/g,
            replacement: function (match, p1, offset, string) {
              return TOKENS[p1];
            }
          }]
        })
      },
      /* Embed files. */
      {
        test: /\.(html|css)$/,
        loader: 'raw-loader',
        exclude: /\.async\.(html|css)$/
      },
      /* Async loading. */
      {
        test: /\.async\.(html|css)$/,
        loaders: ['file?name=[name].[hash].[ext]', 'extract']
      }
    ]
  },
  resolve: {
    extensions: ['.ts', '.js'],
    alias: {
      'angular-stormpath$': path.resolve(__dirname, 'src/index.ts')
    }
  },
  devServer: {
    port: 8000,
    inline: true,
    hot: true,
    historyApiFallback: true,
    contentBase: 'demo',
    proxy: [
      {
        context: [
          '/forgot',
          '/login',
          '/logout',
          '/me',
          '/oauth',
          '/register'
        ],
        target: 'http://localhost:3000',
        secure: false
      }
    ]
  },
  plugins: [
    new TsConfigPathsPlugin(),
    new webpack.ContextReplacementPlugin(
      // The (\\|\/) piece accounts for path separators in *nix and Windows
      /angular(\\|\/)core(\\|\/)(esm(\\|\/)src|src)(\\|\/)linker/,
      root('./src') // location of your src
    ),
    new webpack.DefinePlugin({
      ENV: JSON.stringify(IS_PROD ? 'production' : 'development')
    }),
    new webpack.HotModuleReplacementPlugin(),
    // the LoaderOptionsPlugin is necessary to configure tslint, but causes issues with StringReplacePlugin
    // https://github.com/wbuchwalter/tslint-loader/issues/38
    /*new LoaderOptionsPlugin({
      options: {
        tslint: {
          emitErrors: false,
          failOnHint: false
        }
      }
    }),*/
    new StringReplacePlugin()
  ]
};

function root(__path) {
  return path.join(__dirname, __path);
}
